### ISSUE TITLE
INSERT ON CONFLICT doesn't work for conflicts in same statement

### ISSUE THREAD

**Issue by CGamesPlay** [duckdb/duckdb#8147]
### What happens?

`ON CONFLICT` added support for transactions in #7407, however it does not support conflicts within a single statement. Two reproductions are included. The use case here is creating an aggregation over a large file where I can't use a `GROUP BY` because the aggregation doesn't fit into main memory.

### To Reproduce

```
DuckDB Web Shell
Database: v0.8.2-dev8
Package:  @duckdb/duckdb-wasm@1.27.1-dev3.0

Connected to a local transient in-memory database.
Enter .help for usage hints.

duckdb> create table input (number INTEGER, str VARCHAR);
...

duckdb> INSERT INTO input VALUES (1, 'a'), (2, 'b'), (1, 'c');
...

duckdb> create table output (number INTEGER PRIMARY KEY, str VARCHAR);
...

duckdb> INSERT INTO output SELECT * FROM input ON CONFLICT DO NOTHING;
Constraint Error: PRIMARY KEY or UNIQUE constraint violated: duplicate key "1"

duckdb> insert into output VALUES (1, 'a'), (1, 'b') ON CONFLICT DO NOTHING;
Constraint Error: PRIMARY KEY or UNIQUE constraint violated: duplicate key "1"
```

### OS:

DuckDB Web Shell

### DuckDB Version:

v0.8.2-dev8

### DuckDB Client:

Web Shell

### Full Name:

Ryan Patterson

### Affiliation:

BigData Boutique

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Tishj** [comment]
Thanks for your interest in the feature,

This is currently an expected limitation however, I recommend using DISTINCT to get rid of the conflicts within the inserted data.

**Comment by CGamesPlay** [comment]
Hi, thanks for the speedy response! When you say "expected limitation", does that mean that there are no plans to fix it?

My specific use case is to parse a large log file which has many duplicated records. I want to create a table which contains deduplicated records along with an occurrence count. Using `INSERT ... ON CONFLICT DO UPDATE SET count = count + 1` seems like an obvious approach, but does not work due to this issue.

An alternative approach using `GROUP BY` is also intractable, possibly due to another bug: using `FIRST`/`ANY_VALUE` balloons the memory usage to multiple times the uncompressed size of my input (e.g. I have a ndjson+gzip file that is 2.4 GiB uncompressed, running `ANY_VALUE(some_field)` results in 8.7 GiB of peak memory usage by DuckDB, even though the resulting database is 54 MiB. I have about 10 fields and don't have 87 GiB of RAM, so I can't use `GROUP BY` with `ANY_VALUE`.

My least-desirable approach is to create the DISTINCT index, then do some sort of join to rescan the file for each primary key and find a matching record. This is untenable for speed reasons.

I think my only remaining workaround is to use SQLite to do the aggregation and then transferring the results to DuckDB afterwards.

**Comment by Tishj** [comment]
That does not mean it's intended to not be supported, but it's expected to fail currently and I don't have a timeline when it will be added.

I'm a little confused why
```sql
INSERT INTO output SELECT DISTINCT ON (number) * FROM input ON CONFLICT DO NOTHING;
```
would not be a viable solution, that's essentially what we'd need to turn the INSERT ON CONFLICT into if we wanted to support this, which I've opted not to do because that would cause unnecessary overhead when it's not required.

**Comment by CGamesPlay** [comment]
Sorry for not being more clear: it's not a viable solution because I don't want `DO NOTHING` conflict resolution. I have specific `DO UPDATE` expressions for most of the fields that I want to properly handle the conflict resolutions.

It is possible to get the result I want using only aggregate functions, however I encountered a bug in `ANY_VALUE` that causes RAM usage to go crazy (for LIST fields? unclear). After playing with it for a few hours, I did discover that I can use other aggregate functions like `list_distinct(list_aggregate(fields.query, 'list'))`, which don't exhibit this RAM explosion and do well for my dataset. So, I do have my workaround now, but it would still be desirable to be able to explicitly list out my `ON CONFLICT DO UPDATE` expressions and have DuckDB respect them.
